text/0323-array-functions.md

```diff
@@ -0,0 +1,130 @@
+
+## Detailed design
+
+For each method in the `MutableArray` mixin we will export a similarly named function from `@ember/array`.
+
+For example, the method `array.objectAt(index)` will map to the function `objectAt(array, index)` with implementation
+
+```js
+function objectAt(array, index) {
+  if (Array.isArray(array)) {
+    return array[index];
+  } else {
+    return array.objectAt(index);
+  }
+}
+```
+
+As you can see, the native array case is handled explicitly in order to avoid assuming that the global `Array` prototype has been extended, but otherwise we just delegate to the `objectAt` method.
+
+For reference, below is an expandable list of all array functions we will need.
+
+<details>
+  <summary>List of array functions (49)</summary>
+  <code>
+    <ul>
+      <li>addArrayObserver</li>
+      <li>addObject</li>
+      <li>addObjects</li>
+      <li>any</li>
+      <li>arrayContentDidChange</li>
+      <li>arrayContentWillChange</li>
+      <li>clear</li>
+      <li>compact</li>
+      <li>every</li>
+      <li>filter</li>
+      <li>filterBy</li>
+      <li>find</li>
+      <li>findBy</li>
+      <li>forEach</li>
+      <li>getEach</li>
+      <li>includes</li>
+      <li>indexOf</li>
+      <li>insertAt</li>
+      <li>invoke</li>
+      <li>isAny</li>
+      <li>isEvery</li>
+      <li>lastIndexOf</li>
+      <li>map</li>
+      <li>mapBy</li>
+      <li>objectAt</li>
+      <li>objectsAt</li>
+      <li>popObject</li>
+      <li>pushObject</li>
+      <li>pushObjects</li>
+      <li>reduce</li>
+      <li>reject</li>
+      <li>rejectBy</li>
+      <li>removeArrayObserver</li>
+      <li>removeAt</li>
+      <li>removeObject</li>
+      <li>removeObjects</li>
+      <li>replace</li>
+      <li>reverseObjects</li>
+      <li>setEach</li>
+      <li>setObjects</li>
+      <li>shiftObject</li>
+      <li>slice</li>
+      <li>sortBy</li>
+      <li>toArray</li>
+      <li>uniq</li>
+      <li>uniqBy</li>
+      <li>unshiftObject</li>
+      <li>unshiftObjects</li>
+      <li>without</li>
+    </ul>
+  </code>
+</details>
+
+## How we teach this
+
+This functions must obviously be documented under the `@ember/array` package in the API docs.
+
+We should encourage people to use the phrase "array functions" to describe these functions and "array methods" to describe the method versions.
+
+We should encourage their usage in addons and fastboot apps instead of using `Ember.A`.
+
+In the future, we may want to encourage their usage in all apps so that we can deprecate extending the global `Array` prototype, but that is out of scope for this RFC.
+
+## Drawbacks
+
+This RFC proposes a considerable increase in API surface area however it's mostly duplicating an existing, well known API so it does not increase the mental burden very much.
+
+I also expect the total number of bytes added to the framework be fairly small since it is mostly just shuffling around existing code.
+
+## Alternatives
```

Comment:
The RFC does not mention why we have those functions at all. One could imagine that using something like lodash would be another alternative 🤔

Reply:
I can add an explanation. The answer is that we need them for the same reason that we need `get` and `set`, however unlike `get` and `set` there's no easy way to intercept array indexing with getters/setters without using `Proxy`.

It's possible to remove the need for them entirely by making dramatic changes to the Ember programming model (e.g. moving to a more React/Glimmer.js programming model) but that's out of scope for this RFC.

Reply:
Another alternative is to _only_ expose `objectAt` and `replace` functions (because all the other methods are based on those). This would mean, though, that something as simple as `array.pushObject(item)` would need to be written as `replace(array, get(array, 'length'), 0, [item])`